tools/node_maintain.py

```diff
@@ -0,0 +1,220 @@
+        if state in {"DECOMMISSIONING", "DECOMMISSIONED"} and node not in decommissioned_nodes:
+            unready_nodes[node] = state
+    return unready_nodes
+
+
+def get_decommission_nodes(args):
+    k8s_operator = KubernetesOperator(args.ip)
+    existing_nodes = k8s_operator.get_nodes()
+    logger.info("Current unhealthy node list: {}".format(','.join(existing_nodes)))
+    return existing_nodes
+
+
+def add_decommission_nodes(args):
+    k8s_operator = KubernetesOperator(args.ip)
+    existing_nodes = k8s_operator.get_nodes()
+    nodes = args.nodes
+    if isinstance(nodes, str):
+        nodes = set(nodes.split(','))
+    full_list = existing_nodes | nodes
+    k8s_operator.set_nodes(full_list)
+    logger.info("Add node: {}".format(args.nodes))
+    return full_list
+
+
+def remove_decommission_nodes(args):
+    k8s_operator = KubernetesOperator(args.ip)
+    existing_nodes = k8s_operator.get_nodes()
+    nodes = args.nodes
+    if isinstance(nodes, str):
+        nodes = set(nodes.split(','))
+    full_list = existing_nodes - nodes
+    k8s_operator.set_nodes(full_list)
+    logger.info("Remove node: {}".format(args.nodes))
+    return full_list
+
+
+def update_decommission_nodes(args):
+    k8s_operator = KubernetesOperator(args.ip)
+    nodes = args.nodes
+    if isinstance(nodes, str):
+        nodes = set(nodes.split(','))
+    k8s_operator.set_nodes(nodes)
+    logger.info("Update node list: {}".format(args.nodes))
+    return nodes
+
+
+def refresh_yarn_nodes(args):
+    k8s_operator = KubernetesOperator(args.ip)
+    yarn_operator = YarnOperator(args.ip)
+    while True:
+        yarn_operator.decommission_nodes()
+        current_status = yarn_operator.get_node_status()
+        decommissioned_nodes = k8s_operator.get_nodes()
+        unready_nodes = get_unready_nodes(decommissioned_nodes, current_status)
+        if len(unready_nodes) == 0:
+            break
+        else:
+            logger.info("Unready nodes: {}".format(unready_nodes))
+        logger.info("Waiting...")
+        time.sleep(30)
+    logger.info("Successfully refresh nodes.")
+
+def setup_parser():
+    top_parser = argparse.ArgumentParser()
+    top_parser.add_argument("ip", help="Master node ip")
+    subparser = top_parser.add_subparsers(
+        title="subcommands",
+        description="valid subcommands",
+        dest="subcommands")
+
+    # node list parser
+    nodelist_parser = subparser.add_parser("node-list")
+    nodelist_subparsers = nodelist_parser.add_subparsers(dest="action")
+
+    parser_get = nodelist_subparsers.add_parser("get", help="get unhealthy node list")
+    parser_get.set_defaults(func=get_decommission_nodes)
+
+    parser_add = nodelist_subparsers.add_parser("add", help="add unhealthy nodes")
+    parser_add.add_argument("nodes", help='support comma-delimited node list')
+    parser_add.set_defaults(func=add_decommission_nodes)
+
+    parser_remove = nodelist_subparsers.add_parser("remove", help="remove unhealthy nodes")
```

Comment:
"remove" means decommission? better change it to "decommission", in help, you better say "graceful decommissioning nodes, jobs on the nodes will not be killed."

Reply:
Nope, refined parser helper